File: token-client-core/src/main/java/no/nav/security/token/support/client/core/auth/ClientAssertion.java
package no.nav.security.token.support.client.core.auth;

import com.nimbusds.jose.JOSEException;
import com.nimbusds.jose.JOSEObjectType;
import com.nimbusds.jose.JWSAlgorithm;
import com.nimbusds.jose.JWSHeader;
import com.nimbusds.jose.crypto.RSASSASigner;
import com.nimbusds.jose.jwk.RSAKey;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.SignedJWT;
import jakarta.validation.constraints.NotNull;
import no.nav.security.token.support.client.core.ClientAuthenticationProperties;

import java.net.URI;
import java.sql.Date;
import java.time.Instant;
import java.util.UUID;

public class ClientAssertion {

    private static final int EXPIRY_IN_SECONDS = 60;
    private final URI tokenEndpointUrl;
    private final String clientId;
    private final RSAKey rsaKey;
    private final int expiryInSeconds;

    public ClientAssertion(@NotNull URI tokenEndpointUrl,
                           @NotNull ClientAuthenticationProperties clientAuthenticationProperties) {
        this(
            tokenEndpointUrl,
            clientAuthenticationProperties.getClientId(),
            clientAuthenticationProperties.getClientRsaKey(),
            EXPIRY_IN_SECONDS
        );
    }

    public ClientAssertion(URI tokenEndpointUrl, String clientId, RSAKey rsaKey, int expiryInSeconds) {
        this.tokenEndpointUrl = tokenEndpointUrl;
        this.rsaKey = rsaKey;
        this.clientId = clientId;
        this.expiryInSeconds = expiryInSeconds;
    }

    public String assertion() {
        var now = Instant.now();
        return createSignedJWT(rsaKey, new JWTClaimsSet.Builder()
            .audience(tokenEndpointUrl.toString())
            .expirationTime(Date.from(now.plusSeconds(expiryInSeconds)))
            .issuer(clientId)
            .subject(clientId)
            .claim("jti", UUID.randomUUID().toString())
            .notBeforeTime(Date.from(now))
            .issueTime(Date.from(now))
            .build()).serialize();
    }

    public String assertionType() {
        return "urn:ietf:params:oauth:client-assertion-type:jwt-bearer";
    }

    private SignedJWT createSignedJWT(RSAKey rsaJwk, JWTClaimsSet claimsSet) {
        try {
            var header = new JWSHeader.Builder(JWSAlgorithm.RS256)
                .keyID(rsaJwk.getKeyID())
                .type(JOSEObjectType.JWT);
            var signedJWT = new SignedJWT(header.build(), claimsSet);
            signedJWT.sign(new RSASSASigner(rsaJwk.toPrivateKey()));
            return signedJWT;
        } catch (JOSEException e) {
            throw new RuntimeException(e);
        }
    }
}

File: token-client-core/src/main/kotlin/no/nav/security/token/support/client/core/auth/ClientAssertion.kt
package no.nav.security.token.support.client.core.auth

import com.nimbusds.jose.JOSEObjectType.JWT
import com.nimbusds.jose.JWSAlgorithm.RS256
import com.nimbusds.jose.JWSHeader
import com.nimbusds.jose.crypto.RSASSASigner
import com.nimbusds.jose.jwk.RSAKey
import com.nimbusds.jwt.JWTClaimNames.JWT_ID
import com.nimbusds.jwt.JWTClaimsSet
import com.nimbusds.jwt.JWTClaimsSet.Builder
import com.nimbusds.jwt.SignedJWT
import com.nimbusds.oauth2.sdk.auth.JWTAuthentication.CLIENT_ASSERTION_TYPE
import java.net.URI
import java.time.Instant.now
import java.util.*
import no.nav.security.token.support.client.core.ClientAuthenticationProperties
import kotlin.DeprecationLevel.WARNING

class ClientAssertion(private val tokenEndpointUrl : URI, private val clientId : String, private val rsaKey : RSAKey, private val expiryInSeconds : Int) {
    constructor(tokenEndpointUrl: URI, auth :  ClientAuthenticationProperties) : this(tokenEndpointUrl, auth.clientId, auth.clientRsaKey!!, EXPIRY_IN_SECONDS)

    fun assertion()  =
        now().run {
            createSignedJWT(rsaKey, Builder()
                .audience("$tokenEndpointUrl")
                .expirationTime(Date.from(plusSeconds(expiryInSeconds.toLong())))
                .issuer(clientId)
                .subject(clientId)
                .claim(JWT_ID, "${UUID.randomUUID()}")
                .notBeforeTime(Date.from(this))
                .issueTime(Date.from(this))
                .build()).serialize()
        }

    @Deprecated("Use com.nimbusds.oauth2.sdk.auth.JWTAuthentication instead", ReplaceWith("JWTAuthentication.CLIENT_ASSERTION_TYPE"), WARNING)
    fun assertionType() = CLIENT_ASSERTION_TYPE

    private fun createSignedJWT(rsaJwk : RSAKey, claimsSet : JWTClaimsSet) =
        runCatching {
            SignedJWT(JWSHeader.Builder(RS256)
                .keyID(rsaJwk.keyID)
                .type(JWT).build(), claimsSet).apply {
                    sign(RSASSASigner(rsaJwk.toPrivateKey()))
                }
        }.getOrElse {
            throw RuntimeException(it)
        }

    companion object {
        private const val EXPIRY_IN_SECONDS = 60
    }
}